File: solutions/python/bob/1/bob.py
def response(hey_bob):
    if len(hey_bob.strip()) > 0 and hey_bob.strip()[-1] == '?':
        if hey_bob.isupper():
            return "Calm down, I know what I'm doing!"
        return "Sure."
    if hey_bob.isupper():
        return "Whoa, chill out!"
    if hey_bob.isspace() or len(hey_bob) == 0:
        return "Fine. Be that way!"
    return "Whatever."

File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isAlpha, isSpace)

responseFor :: String -> String
responseFor xs
    | isSilence xs = "Fine. Be that way!"
    | isYelling xs && isQuestion xs = "Calm down, I know what I'm doing!"
    | isYelling xs = "Whoa, chill out!"
    | isQuestion xs = "Sure."
    | otherwise = "Whatever."

isYelling :: String -> Bool
isYelling str = any isUpper alphabeticChars && all isUpper alphabeticChars
    where 
        alphabeticChars = filter isAlpha str

isQuestion :: String -> Bool
isQuestion str = not (null str) && last (reverse $ dropWhile isSpace $ reverse str) == '?'

isSilence :: String -> Bool
isSilence str = all isSpace str
